[JENKINS-55048] Add support for plugin Java requirement metadata

daniel-beck opened the pull request:
See [JENKINS-55048](https://issues.jenkins-ci.org/browse/JENKINS-55048).

Downstream of https://github.com/jenkinsci/maven-hpi-plugin/pull/75 and https://github.com/jenkins-infra/backend-update-center2/pull/164.

Future enhancement might be to send the Java version with the request (or consider it already sent by the user agent header) and handle that on the server side, similar to the tiered update sites. But for now, a big red (newly bold) warning should at least get us started.

Here is how it renders:
![image](https://user-images.githubusercontent.com/223853/50774306-5cec6700-1293-11e9-9d84-ae6db0a601cf.png)

(I used http://updates.jenkins.io/temporary-experimental-java11/ for testing.
`Pipeline: Supporting APIs`, aka `workflow-support` is the only one requiring Java 11+.
The others are red because they depend on it directly or transitively.
I used the following system groovy quick&dirty script to check these things:

```groovy
def showDeps(plugin, level) {
  println ("  ".multiply(level)+"${plugin.name}:${plugin.version}")
  
  plugin.getNeededDependencies().each { dep ->
    def dep1 = dep
    showDeps(dep1, level + 1);
  }
}
showDeps(Jenkins.instance.updateCenter.getPlugin('throttle-concurrents'), 0)
```
).

### Proposed changelog entries

* Add support for plugins declaring a minimum Java version in manifest and update center, showing warnings and refusing to load plugins with unsatisfied dependencies.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

### Desired reviewers

@jenkinsci/code-reviewers

daniel-beck committed:
[JENKINS-20679] Add support for plugin Java requirement metadata

daniel-beck committed:
Rename 'minimumJavaVersion' to 'requiredJava'

oleg-nenashev commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -432,6 +432,20 @@ private String getVersionOf(Manifest manifest) {
     /**
+     * Returns the required Java version of this plugin.
+     * @return the required Java version of this plugin.
+     *
```
Document what `null` means here? Probably it also makes sense to extrapolate the core dependency to a Java version if undefined (or Just return core Java version once/if #2949 is integrated)

jglick replied:
I think null is fine here: means the plugin does not specify a version.

batmat replied:
I think so too. And it's consistent with other similar methods here.

oleg-nenashev commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1079,6 +1088,19 @@ public boolean isForNewerHudson() {
+        public boolean isForNewerJava() {
+            try {
+                return requiredJava != null && new VersionNumber(requiredJava).isNewerThan(
+                        new VersionNumber(System.getProperty("java.version")));
```
Needs a comment like above. I would probably move it to a static method in order to integrate with #2949

jglick replied:
as above, this is the wrong property

oleg-nenashev commented on core/src/main/resources/hudson/PluginManager/_table.css:
```diff
@@ -8,3 +8,6 @@
 }
+.compatWarning, .securityWarning {
+    font-weight: bold;
+}
```
line break

oleg-nenashev reviewed (commented):
Generally the PR looks good, added few comments.

I would also consider adding `Maximum-Java-Version` as a part of this story while you are around. E.g. I barely believe that Maven Integration Plugin can be made Java9-compatible easily (CC @aheritier)

daniel-beck committed:
Adapt tests to changed localizable strings

daniel-beck committed:
Update Javadocs and comments

daniel-beck committed:
Add Java version check to setup wizard and installNecessaryPlugins

I'm not sure what the log message is supposed to be for, but other
checks also log, so apply the same behavior for Java dependency.

jglick commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -432,6 +432,20 @@ private String getVersionOf(Manifest manifest) {
+        String v = manifest.getMainAttributes().getValue("Minimum-Java-Version");
+        if (v!= null) return v;
+
+        return null;
```
More simply,

```java
return manifest.getMainAttributes().getValue("Minimum-Java-Version");
```

jglick commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -574,6 +588,15 @@ public boolean hasLicensesXml() {
+            String requiredJavaVersion = getRequiredJavaVersion();
+            if (requiredJavaVersion != null) {
+                // TODO replace with calls to Runtime.version() once we're on Java 9
+                VersionNumber actualVersion = new VersionNumber(System.getProperty("java.version"));
```
You want `java.specification.version`. The format of `java.version` is undefined.

jglick commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1079,6 +1088,22 @@ public boolean isForNewerHudson() {
+                return requiredJava != null && new VersionNumber(requiredJava).isNewerThan(
+                        new VersionNumber(System.getProperty("java.version")));
+            } catch (NumberFormatException nfe) {
+                return false; // plugin doesn't declare a minimum Java version
```
Huh? This should be logged as a `WARNING`.

daniel-beck replied:
I really don't see how this could happen, but basically copied it from the existing code, which behaves identically.

batmat replied:
I agree with @jglick this looks wrong. The comment is misleading: this will enter the catch block iff `minimumJavaVersion` is not null, and hence specified.

Then would enter the block if `new VersionNumber(minimumJavaVersion)` throws a `NumberFormatException`. So if a minimumJavaVersion **is** set AND cannot be parsed: this should very likely be reported or I'm missing something.

jglick commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1093,6 +1118,32 @@ public VersionNumber getNeededDependenciesRequiredCore() {
+            try {
+                versionNumber = requiredJava == null ? null : new VersionNumber(requiredJava);
+            } catch (NumberFormatException nfe) {
+                // unable to parse version
```
log

jglick reviewed (dismissed):
Misuse of `java.version`, otherwise looks good.

daniel-beck commented:
Rebased to current master with no impl changes so far.

daniel-beck committed:
Use java.specification.version

daniel-beck committed:
Adapt to newer core internals

batmat committed:
Merge branch 'master' into JENKINS-20679-filter-out-plugins-compiled-with-more-recent-jdk

batmat committed:
Rename to minimumJavaVersion to keep the name across tooling

Given this field is present in maven-hpi-plugin, update-center, etc.
I think it does not hurt to keep the 'Minimum-Java-Version' name across
various usages to help make it clear.

batmat committed:
Add warning if minimumJavaVersion is not parseable

batmat committed:
Add test covering methods for "is*ForNewerJava()"

This way at least we know the backend methods should not be broken without
noticing.

batmat committed:
Log warning in case of unparseable minimum Java version

batmat commented on test/src/test/resources/plugins/minJavaVersion-update-center.json:
```diff
@@ -0,0 +1,57 @@
+            "title": "Dummy",
+            "url": "http://nowhere.net/dummy.hpi",
+            "version": "1.0",
+            "minimumJavaVersion": "25"
```
Hopefully, we shouldn't get too quickly to this as a baseline (?). Or maybe I should even have gone bigger.

batmat commented:
@jenkinsci/java11-support please review this pull-request.

batmat commented:
To test this, a simple way is the following:

```shell
# Build and run a Jenkins with this patch
docker run -ti -e ID=3016 -p 8080:8080 batmat/jenkins-pr-tester
```

Then open a browser on 8080, and change the update center URL to http://updates.jenkins.io/temporary-experimental-java11/update-center.json to check how it behaves. 
In this UC, the `workflow-support` plugin in 3.0-java11-alpha-11` requires Java 11.

batmat commented:
Tested also using the `jenkins/jenkins:jdk11` Docker image, the warnings are not shown, as expected.

`docker run -ti -p 8080:8080 -v ~/path/to/repo/jenkinsci/jenkins/war/target/jenkins.war:/usr/share/jenkins/jenkins.war:ro jenkins/jenkins:jdk11`

MarkEWaite commented on test/src/test/resources/plugins/minJavaVersion-update-center.json:
```diff
@@ -0,0 +1,57 @@
+updateCenter.post(
+{"connectionCheckUrl":"http://www.google.com/", // has to be here because updateDirectly sniffs only very limited formats
```
JSON doesn't allow comments.  Does this work?

If so, why?

batmat replied:
> Does this work?

Yes. It works, you can see the test does use it.

> If so, why?

It's history I suppose :). This is the format used on all Jenkins update center files. 
You can check for yourself in `curl -L -r 99 https://updates.jenkins.io/update-center.json`

(BTW, a few years ago were added the actual json exports, add `.actual` in between, like https://updates.jenkins.io/update-center.actual.json).

alecharp replied:
because it's not  JSON but JSONP, maybe comments are authorized in this format.

fcojfernandez replied:
AFAIK, neither JSON nor JSONP allow comment, but depending on the library you use for parsing the data, the comments are ignored.

MarkEWaite replied:
Would you be willing to remove the comment, or place it in an unused data element?  

I think GitHub is right to flag it with red in the diffs, since the json file format does not allow comments.

If not, I'm ok with that as well.

batmat replied:
I'm going to remove it. I think it works because the parser is permissive.

For the record, I didn't write this myself. There are already many similar `.json` files in that resource directory which I copied and modified for my test case.

batmat replied:
Done @MarkEWaite

daspilker commented:
Pinging @sghill as the new maintainer of the Gradle JPI plugin.

MRamonLeon commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1118,6 +1126,24 @@ public boolean isForNewerHudson() {
         }
 
+        /**
+         * Returns true iff the plugin declares a minimum Java version and it's newer than what the Jenkins master is running on.
```
typo: iff -> if

MRamonLeon replied:
gotcha

MRamonLeon commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1118,6 +1126,24 @@ public boolean isForNewerHudson() {
+        /**
+         * Returns true iff the plugin declares a minimum Java version and it's newer than what the Jenkins master is running on.
+         *
+         * @return true iff the plugin declares a minimum Java version and it's newer than what the Jenkins master is running on.
```
same typo

MRamonLeon commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1118,6 +1126,24 @@ public boolean isForNewerHudson() {
+                return minimumJavaVersion != null && new VersionNumber(minimumJavaVersion).isNewerThan(
+                        new VersionNumber(System.getProperty("java.specification.version")));
+            } catch (NumberFormatException nfe) {
+                logBadMinJavaVersion("'minimumJavaVersion' was specified for plugin ''{0}'' but unparseable (received {1})", this.name,
```
Can you set the message inside the method avoiding passing it as an argument?. It's always the same

batmat replied:
Yeah, I considered it, but didn't do it because this seemed not like a big deal. Looking.

MRamonLeon commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1148,6 +1206,22 @@ public boolean isNeededDependenciesForNewerJenkins(PluginManager.MetadataCache c
         }
 
+        /**
+         * Returns true iff this plugin or any of its dependencies require a newer Java than Jenkins is running on.
```
:laughing:  look for *iff* in code

MRamonLeon commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1148,6 +1206,22 @@ public boolean isNeededDependenciesForNewerJenkins(PluginManager.MetadataCache c
+        /**
+         * Returns true iff this plugin or any of its dependencies require a newer Java than Jenkins is running on.
+         *
+         * @return true iff this plugin or any of its dependencies require a newer Java than Jenkins is running on.
```
The comment doesn't fit the code, I think. The code is not taking into account if this plugin *isForNewerJava*, it only checks its dependencies. Or I'm missing something. I don't know whether the code is wrong, or just the comment

batmat replied:
fixed, nice catch.

MRamonLeon commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1132,6 +1158,38 @@ public VersionNumber getNeededDependenciesRequiredCore() {
+         * @since TODO
+         */
+        @CheckForNull
+        public VersionNumber getNeededDependenciesMinimumJavaVersion() {
```
The name of the method is tricky for me. Do you like instead, something like **getMinimumJavaVersionConsideringDependencies** or similar? :pray:

batmat replied:
I tend to agree, but I used the existing naming pattern. So I think either we'd fix everything, or nothing.

MRamonLeon replied:
Nothing to be consistent then.

batmat replied:
In *this* PR, I agree adjusting everything would be out of scope. But fixing it in a dedicated PR could make sense.

batmat committed:
Fix Javadoc: isForNewerJava() is the one checking the current plugin

batmat committed:
Factor out the message and plugin name in submethod

alecharp commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1118,6 +1126,23 @@ public boolean isForNewerHudson() {
+        public boolean isForNewerJava() {
+            try {
+                return minimumJavaVersion != null && new VersionNumber(minimumJavaVersion).isNewerThan(
+                        new VersionNumber(System.getProperty("java.specification.version")));
```
between this and [#763](https://github.com/jenkinsci/jenkins/pull/3016/files#diff-dd00e0527600a2b32c10fe2d5af188adR763), could we externalize (in a method) how we get the Java version? This way, we make sure that any part of the code is using the same system and when we will be able to change it by `Runtime.version()` we will have to do it in one place?

batmat replied:
Yes, absolutely. I almost did it, but didn't want to pull myself into an out of scope change in the same PR. (_looks_ tiny, granted, but devil's in the details...). I had a look, but given Runtime.version() returns a new `Runtime.Version` class, it would not be _such_ a straightforward and forward compatible function.

I would propose to file a separate PR to group the retrievals to java.specification.version (or wrong java.version) as there are many in the core.
Likely, this could even be centralized in the versionnumber-lib.

jglick replied:
At least a method in `Util` for now.

batmat replied:
done

MRamonLeon commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1207,9 +1206,9 @@ public boolean isNeededDependenciesForNewerJenkins(PluginManager.MetadataCache c
 
         /**
-         * Returns true iff this plugin or any of its dependencies require a newer Java than Jenkins is running on.
+         * Returns true iff any of it the plugin dependencies require a newer Java than Jenkins is running on.
```
I think you can omit the word **it** here and in the @return section

MRamonLeon commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1186,8 +1184,9 @@ public VersionNumber getNeededDependenciesMinimumJavaVersion() {
 
-        private void logBadMinJavaVersion(String s, String pluginName, String minimumJavaVersion) {
-            LOGGER.log(Level.WARNING, s, new String[]{pluginName, minimumJavaVersion});
+        private void logBadMinJavaVersion(String minimumJavaVersion) {
```
As the calls to this method is always the same you can even avoid passing the **minimumJavaVersion** argument.

batmat replied:
:+1: done.

MRamonLeon reviewed (approved):
Minor comments, but anyway it LGTM

batmat committed:
Remove json comment

jglick commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -495,6 +495,17 @@ private String getVersionOf(Manifest manifest) {
     }
 
+    /**
+     * Returns the minimumJava version of this plugin.
```
missing space

jglick commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -495,6 +495,17 @@ private String getVersionOf(Manifest manifest) {
 
+    /**
+     * Returns the minimumJava version of this plugin.
+     * @return the minimum Java version of this plugin.
```
Please specify the format here. `1.8`? `8`? Something else?

jglick commented on core/src/main/java/hudson/PluginWrapper.java:
```diff
@@ -495,6 +495,17 @@ private String getVersionOf(Manifest manifest) {
+     * Returns the minimumJava version of this plugin.
+     * @return the minimum Java version of this plugin.
+     *
+     * @since TODO
```
BTW the practice has been to use `FIXME` in these comments intended to be cleaned up immediately after the merge if not in the merge commit itself, as opposed to “rainy-day” comments that could sit there for years. I guess to make it easier to grep for them.

batmat replied:
It _seems_ like there are actually `@since TODO` only currently in the codebase.

Anyway, I filed #3837 to address these more easily in the future.

jglick commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1118,6 +1126,23 @@ public boolean isForNewerHudson() {
+        /**
+         * Returns true iff the plugin declares a minimum Java version and it's newer than what the Jenkins master is running on.
+         *
+         * @return true iff the plugin declares a minimum Java version and it's newer than what the Jenkins master is running on.
```
Avoid duplicating text between summary and `@return` tag. You can omit one or the other, or use the `@return` tag to delve into details of format, while the summary gives the overall purpose.

batmat replied:
done

jglick commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1132,6 +1157,38 @@ public VersionNumber getNeededDependenciesRequiredCore() {
+         *
+         * @since TODO
+         */
+        @CheckForNull
```
Meaning of null should be documented in `@return`.

batmat replied:
done

jglick commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1132,6 +1157,38 @@ public VersionNumber getNeededDependenciesRequiredCore() {
+        }
+
+        private void logBadMinJavaVersion(String minimumJavaVersion) {
+            LOGGER.log(Level.WARNING, "'minimumJavaVersion' was specified for plugin ''{0}'' but unparseable (received {1})",
```
Might have missed a doubling of `'` here. I recommend just avoiding ASCII apostrophes in message formats. Use `‘…’` or nothing.

jglick commented on core/src/main/resources/hudson/PluginManager/table.properties:
```diff
@@ -34,6 +37,10 @@ depCoreWarning=\
  Jenkins will refuse to load the dependent plugins requiring a newer version of Jenkins, \
  and in turn loading this plugin will fail.
+depJavaWarning=\
+ Warning: This plugin requires dependent plugins that require Java {0} or newer. \
```
_requires dependent plugins_ reads a little awkwardly—it is redundant. I might suggest something like

> Warning: this plugin requires other plugins which in turn require Java…

jglick commented on test/src/test/java/hudson/model/UpdateSiteTest.java:
```diff
@@ -153,6 +153,31 @@ public void shutdownWebserver() throws Exception {
         assertNotEquals("plugin data is present", Collections.emptyMap(), site.getData().plugins);
     }
 
+    @Test public void minimumJavaVersion() throws Exception {
```
`@Issue`

jglick commented on test/src/test/resources/plugins/minJavaVersion-update-center.json:
```diff
@@ -0,0 +1,57 @@
+updateCenter.post(
```
This is JSONP. Either put it in a file with a `.js` extension, or actually use JSON, which IIRC has been supported for a few years now.

batmat replied:
I don't disagree, it just feels out of scope given there're many files like these already in the folder. But can fix in a followup if you feel any strongly.

batmat committed:
Add missing space in javadoc

batmat committed:
Remove duplicated Javadoc

as proposed by @jglick

batmat committed:
Add reference to related JIRA entry

batmat committed:
Provide more information about where/how the `Minimum-Java-Version` is coming from the plugin's metadata

batmat committed:
Clarify what `null` means

batmat committed:
Remove apostrophes to avoid issues in string un/expansion

batmat committed:
Remove spurious 'it' (and duplicate doc)

batmat committed:
Make the phrasing less awkward

using the proposal from @jglick in https://github.com/jenkinsci/jenkins/pull/3016/files#r246146054

batmat committed:
Remove also param given minimumJavaVersion is a class member already

batmat committed:
Introduce method to fetch java.specification.version

batmat commented:
@jglick I believe I addressed your feedback. Can you please take another quick look?

Thanks!

batmat commented:
BTW, anyone willing to have another look is welcome.
The additions I made since approvals are very trivial, often in javadoc only but a few confirmed approvals would be better I suppose.

Thanks!

jglick commented on core/src/main/java/jenkins/util/java/JavaUtils.java:
```diff
@@ -51,8 +52,23 @@ public static boolean isRunningWithJava8OrBelow() {
     }
 
+    /**
+     * Returns the JVM's current version as a {@link VersionNumber} instance.
```
Best to document this with examples. IIUC these would be

* `1.8`
* `9`
* `10`
* `11`

?

batmat replied:
Yes. Basically [JEP-223](https://openjdk.java.net/jeps/223)

batmat replied:
@jglick fix proposed in https://github.com/jenkinsci/jenkins/pull/3842

oleg-nenashev commented:
It causes https://issues.jenkins-ci.org/browse/JENKINS-55562 , because there is no version normalization code in this pull request. Both `8` and `1.8` are valid version numbers, because pre-JEP-223 format stays valid for older versions
